arb/nth_root_pol.c

```diff
@@ -49,11 +49,78 @@ nth_root_pol_prod(acb_t y, acb_srcptr u, const arb_t x, slong d, slong m, slong
 void
 nth_root_pol_turn(acb_t y, acb_srcptr u, const arb_t x, acb_srcptr z, slong d, slong m, slong prec)
 {
-    slong q; /* integer mod 2m */
+  
+    slong q, k, sgn_res;  
+    acb_t s, t;
+    arb_t isgn_s; 
+    arb_t isgn_t;
 
-    /* make and product and update q */
+    acb_init(s);
+    acb_init(t);
+    arb_init(isgn_s);
+    arb_init(isgn_t);
+    q = 0;
+    acb_one(y);
+    acb_set_arb(s, x);
+    acb_sub(s, s, u + 0, prec);
+
+    if ( acb_is_real(s) && arb_is_negative(acb_realref(s)) )
+    {      
+        q++;
+        acb_mul_si(s,s,-1,prec);
+    }
+    arb_sgn(isgn_s,acb_imagref(s));
+
+    for (k = 1; k < d; k++)
+    { 
+      acb_set_arb(t, x);
+      acb_sub(t, t, u + k, prec);
 
-    /* multiply by (z^1/2)^q */
-    if (q)
-        acb_mul(y, y, z + q, prec);
+      if ( acb_is_real(t) && arb_is_negative(acb_realref(t)) )
+      {
+        sgn_res++;
+        q++;
+        acb_mul_si(t,t,-1,prec);
+      }
+      arb_sgn(isgn_t,acb_imagref(t));
+      acb_mul(s, s, t, prec);
+      if ( arb_eq(isgn_s,isgn_t) )
+      {
+        arb_sgn(isgn_s,acb_imagref(s));
+        if ( !arb_overlaps(isgn_s,isgn_t) )
+        {
+          if ( arb_contains_nonnegative(isgn_t) )
+          {
+            q = q + 2;
+          }
+          else
+          {
+            q = q - 2;
+          }
+        }
+      }
+      else
+      {
+         arb_sgn(isgn_s,acb_imagref(s));
+      }
+
+      if ( acb_is_real(s) && arb_is_negative(acb_realref(s)) )
+      {
+        sgn_res++;
+        q++;
+        acb_mul_si(s,s,-1,prec);
+      }
+    } 
+
+    acb_root_ui(s, s, m, prec);
```

Comment:
the point of having an argument acb_srcptr is to have all roots zeta^(i/2m) precomputed, otherwise you gain nothing. So you should just multiply by z + q if q is non zero